File: solutions/__125_is_cycle_list2.go
package main

// ListNode for singly-linked list.
type ListNode struct {
	Val  int
	Next *ListNode
}

// O(n) O(n) 时间复杂度
func detectCycle(head *ListNode) *ListNode {
	hashmap := map[*ListNode]bool{}
	for head != nil {
		if hashmap[head] {
			return head
		}
		hashmap[head] = true
		head = head.Next
	}
	return nil
}

File: solutions/__125_is_cycle_list2.cpp
//
// Created by peanut on 2021/8/17.
//

#include <bits/stdc++.h>

using namespace std;

struct ListNode {
    int val;
    ListNode *next;

    ListNode(int x) : val(x), next(NULL) {}
};

ListNode* detectCycle(ListNode *head) {
    ListNode *slow = head;
    ListNode *fast = slow;
    while (fast != nullptr) {
        slow = slow->next;
        if (fast->next == nullptr)
            return nullptr;
        fast = fast->next->next;
        if (slow == fast) {
            while (head != slow) {
                head = head->next;
                slow = slow->next;
            }
            return head;
        }
    }
    return nullptr;
}